File: FlightBooking.Infrastructure/Services/AircraftService.cs
using FlightBooking.Application.Common.Exceptions;
using FlightBooking.Application.Features.Flights.DTOs;
using FlightBooking.Application.Features.Flights.Interfaces;
using FlightBooking.Domain.Entities.Flights;
using FlightBooking.Infrastructure.Persistence;
using Microsoft.EntityFrameworkCore;

namespace FlightBooking.Infrastructure.Services
{
    public class AircraftService : IAircraftService
    {
        private readonly FlightBookingDbContext _context;

        public AircraftService(FlightBookingDbContext context)
        {
            _context = context;
        }

        private static AircraftDto MapToDto(Aircraft a) => new()
        {
            Id = a.Id,
            RegistrationNumber = a.RegistrationNumber,
            Model = a.Model,
            TotalSeats = a.TotalSeats,
            AirlineId = a.AirlineId,
            AirlineName = a.Airline?.Name ?? string.Empty,
            IsActive = a.IsActive
        };

        public async Task<List<AircraftDto>> GetAllAsync()
        {
            return await _context.Aircrafts
                .Include(a => a.Airline)
                .Select(a => new AircraftDto
                {
                    Id = a.Id,
                    RegistrationNumber = a.RegistrationNumber,
                    Model = a.Model,
                    TotalSeats = a.TotalSeats,
                    AirlineId = a.AirlineId,
                    AirlineName = a.Airline!.Name,
                    IsActive = a.IsActive
                }).ToListAsync();
        }

        public async Task<AircraftDto> GetByIdAsync(int id)
        {
            var aircraft = await _context.Aircrafts
                .Include(a => a.Airline)
                .FirstOrDefaultAsync(a => a.Id == id)
                ?? throw new NotFoundException("Aircraft", id);
            return MapToDto(aircraft);
        }

        public async Task<List<AircraftDto>> GetByAirlineAsync(int airlineId)
        {
            return await _context.Aircrafts
                .Include(a => a.Airline)
                .Where(a => a.AirlineId == airlineId)
                .Select(a => new AircraftDto
                {
                    Id = a.Id,
                    RegistrationNumber = a.RegistrationNumber,
                    Model = a.Model,
                    TotalSeats = a.TotalSeats,
                    AirlineId = a.AirlineId,
                    AirlineName = a.Airline!.Name,
                    IsActive = a.IsActive
                }).ToListAsync();
        }

        public async Task<AircraftDto> CreateAsync(CreateAircraftRequest request)
        {
            var airlineExists = await _context.Airlines.AnyAsync(a => a.Id == request.AirlineId);
            if (!airlineExists) throw new NotFoundException("Airline", request.AirlineId);

            var duplicate = await _context.Aircrafts
                .AnyAsync(a => a.RegistrationNumber == request.RegistrationNumber);
            if (duplicate) throw new BadRequestException("Số đăng ký tàu bay đã tồn tại.");

            var aircraft = new Aircraft
            {
                RegistrationNumber = request.RegistrationNumber,
                Model = request.Model,
                TotalSeats = request.TotalSeats,
                AirlineId = request.AirlineId,
                IsActive = true
            };
            _context.Aircrafts.Add(aircraft);
            await _context.SaveChangesAsync();
            return await GetByIdAsync(aircraft.Id);
        }

        public async Task<bool> UpdateAsync(int id, UpdateAircraftRequest request)
        {
            var aircraft = await _context.Aircrafts.FindAsync(id)
                ?? throw new NotFoundException("Aircraft", id);
            aircraft.Model = request.Model;
            aircraft.TotalSeats = request.TotalSeats;
            aircraft.IsActive = request.IsActive;
            await _context.SaveChangesAsync();
            return true;
        }

        public async Task<bool> DeleteAsync(int id)
        {
            var aircraft = await _context.Aircrafts.FindAsync(id)
                ?? throw new NotFoundException("Aircraft", id);
            var hasFlights = await _context.Flights.AnyAsync(f => f.AircraftId == id);
            if (hasFlights) throw new BadRequestException("Không thể xóa tàu bay đang có lịch bay.");
            _context.Aircrafts.Remove(aircraft);
            await _context.SaveChangesAsync();
            return true;
        }
    }
}

File: Flightbooking-web/src/services/aircraft.service.ts
import apiClient from './apiClient';

export interface Aircraft {
  id: number;
  registrationNumber: string;
  model: string;
  totalSeats: number;
  airlineId: number;
  airlineName: string;
  isActive: boolean;
}

export interface CreateAircraftRequest {
  registrationNumber: string;
  model: string;
  totalSeats: number;
  airlineId: number;
}

export interface UpdateAircraftRequest {
  model: string;
  totalSeats: number;
  isActive: boolean;
}

export const aircraftService = {
  getAll: async (): Promise<Aircraft[]> => {
    const response = await apiClient.get<{ data: Aircraft[], message: string }>('/Aircraft');
    return response.data.data;
  },

  getById: async (id: number): Promise<Aircraft> => {
    const response = await apiClient.get<{ data: Aircraft, message: string }>(`/Aircraft/${id}`);
    return response.data.data;
  },

  create: async (request: CreateAircraftRequest): Promise<Aircraft> => {
    const response = await apiClient.post<{ data: Aircraft, message: string }>('/Aircraft', request);
    return response.data.data;
  },

  update: async (id: number, request: UpdateAircraftRequest): Promise<Aircraft> => {
    const response = await apiClient.put<{ data: Aircraft, message: string }>(`/Aircraft/${id}`, request);
    return response.data.data;
  },

  delete: async (id: number): Promise<void> => {
    await apiClient.delete(`/Aircraft/${id}`);
  }
};
